### PR TITLE
Implement an initial working version of the modified architecture workflow

### PR DESCRIPTION
**Pull request by OrestisAlpos**
This is a draft PR, meant to provide feedback and suggestions, not to be merged yet.
It contains:
- Implement l1_manager binary that is responsible for handling L1 IPC commands
- Implement subnet_interactor binary that is responsible for handling L_i IPC commands where i > 1
- Implement ipc_subnet_state struct, that exposes an interface for child subnet interaction and holds the subnet state
- Introduce new directory `BTC` which describes the subnet hierarchy and each file holds the particular subnet state
- Update docs

### CODE REVIEW

File: src/ipc_subnet_state.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ValidatorData {
+    pub ip: String,
+    pub collateral: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct IPCSubnetState {
```
Comment (OrestisAlpos) [inline src/ipc_subnet_state.rs]: Let's break this state in two pieces: IpcState, SubnetState.
Everything in this file can go into the IpcState, except for the `create_account()`, `transfer()` and `get_checkpoint()` functions, which operate on the SubnetState.
Then we can have the L1 Manager use only IpcState, and the L2 Validator (Subnet Simulator) use only the SubnetState. I have now described the two entities in Notion in detail, so as to clear any confusion.

File: src/ipc_subnet_state.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            .open(&self.file_path)
+            .expect("Failed to open state file");
+
+        file.write_all(json.as_bytes())
+            .expect("Failed to write state to file");
+
+        return json;
+    }
+
+    pub fn create_account(&mut self, address: &str, initial_balance: u64) {
```
Comment (OrestisAlpos) [inline src/ipc_subnet_state.rs]: This should be part of / used by the Subnet Simulator.

File: src/ipc_subnet_state.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            },
+        );
+        self.save_state();
+        println!(
+            "Account {} created with balance {}",
+            address, initial_balance
+        );
+    }
+
+    pub fn transfer(&mut self, from: &str, to: &str, amount: u64) -> Result<(), String> {
```
Comment (OrestisAlpos) [inline src/ipc_subnet_state.rs]: This should be part of / used by the Subnet Simulator.

File: src/ipc_subnet_state.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            .or_insert(Account { balance: 0 });
+
+        to_account.balance += amount;
+
+        self.save_state();
+        println!("Transfer successful");
+        Ok(())
+    }
+
+    pub fn get_checkpoint(&mut self) -> String {
```
Comment (OrestisAlpos) [inline src/ipc_subnet_state.rs]: This should be part of / used by the Subnet Simulator.

File: src/bin/subnet_interactor.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    pub fn new(state: IPCSubnetState) -> Self {
+        SubnetInteractor { state }
+    }
+
+    pub fn interactive_interface(&mut self) {
+        loop {
+            println!("Select an option:");
+            println!("1. Create account");
+            println!("2. Transfer funds");
+            println!("3. Add child subnet");
```
Comment (OrestisAlpos) [inline src/bin/subnet_interactor.rs]: Let's remove this. We are not interested in creating children of L2 subnets. The only "create child" we care about is offered by the L1 Manager.

File: src/bin/subnet_interactor.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            println!("4. Checkpoint state");
```
Comment (OrestisAlpos) [inline src/bin/subnet_interactor.rs:41]: I would suggest we remove that as well. Let's make the Relayer responsible for checkpoints.

File: BTC/B/B.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (OrestisAlpos) [inline BTC/B/B.json]: This is very nice, let's keep this  structure of storing the IPC state. But, as stated elsewhere, let's remove from here everything that is subnet-internal (accounts and their balance) and everything about subnets of a subnet (we do not care about L3+).

File: src/bin/l1_manager.rs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        io::stdin()
+            .read_line(&mut pk)
+            .expect("Failed to read public key");
+
+        println!("Enter collateral (in satoshis):");
+        io::stdin()
+            .read_line(&mut collateral)
+            .expect("Failed to read collateral");
+
+        println!("Enter subnet url:");
```
Comment (OrestisAlpos) [inline src/bin/l1_manager.rs]: The user should not have to input  this. The URL can be "BTC"/name.

File: src/bin/l1_manager.rs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+        println!("Transaction for create child sent successfully, please wait for confirmation");
+    }
+
+    fn join_child(&self) {
+        let mut ip = String::new();
+        let mut pk = String::new();
+        let mut name = String::new();
+
+        println!("Enter IP address:");
```
Comment (OrestisAlpos) [inline src/bin/l1_manager.rs]:
```suggestion
        println!("Enter your IP address:");
```

File: src/bin/l1_manager.rs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        validator_data.push_str(&format!(":ip={}:", ip));
+
+        let pubkey = bitcoin::secp256k1::PublicKey::from_str(&pk).expect("Invalid public key");
+        let subnet_address =
+            bitcoin_utils::get_address_from_public_key(pubkey, bitcoin_ipc::NETWORK);
+
+        validator_data.push_str(&format!("pk={}:", pk));
+        validator_data.push_str(&format!("name={}", name));
+
+        let ipc_state = IPCState::load_state(format!(
```
Comment (OrestisAlpos) [inline src/bin/l1_manager.rs]: Wouldn't it be better if this function loaded the state in the beginning and then asked the user which of the existing subnets it wants to join (the user can just type the subnet name and get an error if it doesn't exist)?

File: src/bin/l1_manager.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        let mut ip = String::new();
+        let mut pk = String::new();
+        let mut name = String::new();
+
+        println!("Enter IP address:");
+        io::stdin()
+            .read_line(&mut ip)
+            .expect("Failed to read IP address");
+
+        println!("Enter public key:");
```
Comment (OrestisAlpos) [inline src/bin/l1_manager.rs]: Is this  the public key of the validator in the subnet or the subnetPK? Currently it seems that we use it as both. If it is the validator subnet public key, then we should add it to the validator_data, and that's all. If it is used to identify the subnet to join, then we also need to ask the user for its validator public key (which btw will not be a bitcoin key but a format defined by the subnet, so it does not have to be parsed here, just included as a string or bytes).

File: src/main.rs [hunk 12]
```diff
@@ -1,21 +1,3 @@
-use bitcoin::{key::Secp256k1, Amount};
-use bitcoin_ipc::{
-    bitcoin_utils::{get_address_from_private_key, get_private_key},
-    ipc_lib::{create_child, join_child},
-};
-
 fn main() -> Result<(), Box<dyn std::error::Error>> {
```
Comment (OrestisAlpos) [inline src/main.rs]: @jakovmitrovski It looks like we do not need a main function anymore, as all functionality has been implemented in the binaries. If you agree, can you delete it completely?

**Review by jakovmitrovski — APPROVED** [review]

File: src/bin/l1_manager.rs [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    fn join_child(&self) {
+        let mut ip = String::new();
+        let mut pk = String::new();
+        let mut username = String::new();
+
+        let subnets = IPCState::load_all().expect("Failed to load subnets");
+
+        let available_subnets: Vec<&IPCState> = subnets
+            .iter()
+            .filter(|subnet| !subnet.has_required_validators())
```
Comment (themicp) [inline src/bin/l1_manager.rs:75]: Should not filter here, otherwise nobody can join after the minimum number of validators have already joined.

**Review by themicp — CHANGES_REQUESTED** [review]

File: src/bin/l1_manager.rs (hunk 13 above)
Comment (OrestisAlpos) [reply src/bin/l1_manager.rs:75]: Valid  comment, but  let's assume  for now that only one validator will want to join

Comment (themicp) [reply src/bin/l1_manager.rs:75]: We can just remove it since it doesn't serve any purpose, no need to accumulate code debt.

Comment (OrestisAlpos) [reply src/bin/l1_manager.rs:75]: Let's do that @jakovmitrovski . Let's allow users to always join.

**Pull request merged**